itest/litd_custom_channels_test.go

```diff
@@ -3304,3 +3320,220 @@ func runCustomChannelsHtlcForceClose(ctxb context.Context, t *harnessTest,
+		t, net, dave, erin, lntest.OpenChannelParams{
+			Amt:         10_000_000,
+			SatPerVByte: 5,
+		},
+	)
+	defer closeChannelAndAssert(t, net, dave, channelOp, false)
+
+	// This is the only public channel, we need everyone to be aware of it.
+	assertChannelKnown(t.t, charlie, channelOp)
+	assertChannelKnown(t.t, fabia, channelOp)
+
+	universeTap := newTapClient(t.t, zane)
+	charlieTap := newTapClient(t.t, charlie)
+	daveTap := newTapClient(t.t, dave)
+	erinTap := newTapClient(t.t, erin)
+	fabiaTap := newTapClient(t.t, fabia)
+	yaraTap := newTapClient(t.t, yara)
+
+	// Mint an asset on Charlie and sync all nodes to Charlie as the
+	// universe.
+	mintedAssets := itest.MintAssetsConfirmBatch(
+		t.t, t.lndHarness.Miner.Client, charlieTap,
+		[]*mintrpc.MintAssetRequest{
+			{
+				Asset: itestAsset,
+			},
+		},
+	)
+	cents := mintedAssets[0]
+	assetID := cents.AssetGenesis.AssetId
+
+	t.Logf("Minted %d lightning cents, syncing universes...", cents.Amount)
+	syncUniverses(t.t, charlieTap, dave, erin, fabia, yara)
+	t.Logf("Universes synced between all nodes, distributing assets...")
+
+	const (
+		daveFundingAmount = uint64(400_000)
+		erinFundingAmount = uint64(200_000)
+	)
+	charlieFundingAmount := cents.Amount - uint64(2*400_000)
+
+	_, _, chanPointEF := createTestAssetNetwork(
+		t, net, charlieTap, daveTap, erinTap, fabiaTap, yaraTap,
+		universeTap, cents, 400_000, charlieFundingAmount,
+		daveFundingAmount, erinFundingAmount, 0,
+	)
+
+	// Before we start sending out payments, let's make sure each node can
+	// see the other one in the graph and has all required features.
+	require.NoError(t.t, t.lndHarness.AssertNodeKnown(charlie, dave))
+	require.NoError(t.t, t.lndHarness.AssertNodeKnown(dave, charlie))
+	require.NoError(t.t, t.lndHarness.AssertNodeKnown(dave, yara))
+	require.NoError(t.t, t.lndHarness.AssertNodeKnown(yara, dave))
+	require.NoError(t.t, t.lndHarness.AssertNodeKnown(erin, fabia))
+	require.NoError(t.t, t.lndHarness.AssertNodeKnown(fabia, erin))
+	require.NoError(t.t, t.lndHarness.AssertNodeKnown(charlie, erin))
+
+	logBalance(t.t, nodes, assetID, "initial")
+
+	// We now deplete the channel between Erin and Fabia by moving all
+	// assets to Fabia.
+	sendAssetKeySendPayment(
+		t.t, erin, fabia, erinFundingAmount, assetID,
+		fn.None[int64](), lnrpc.Payment_SUCCEEDED,
+		fn.None[lnrpc.PaymentFailureReason](),
+	)
+	logBalance(t.t, nodes, assetID, "after moving assets to Fabia")
+
+	// Test case 1: We cannot keysend more assets from Erin to Fabia.
+	sendAssetKeySendPayment(
+		t.t, erin, fabia, 1, assetID, fn.None[int64](),
+		lnrpc.Payment_FAILED, fn.Some(errNoBalance),
+	)
+
+	// Test case 2: We cannot pay an invoice from Charlie to Fabia.
+	invoiceResp := createAssetInvoice(t.t, erin, fabia, 123, assetID)
+	payInvoiceWithSatoshi(
+		t.t, charlie, invoiceResp, lnrpc.Payment_FAILED,
+		fn.Some(errNoRoute),
+	)
+
+	// Test case 3: We now create an asset buy order for a normal amount of
```

Comment:
could incorporate everything so far into the `testCustomChannelsLiquidityEdgeCases` test?

not sure if we rather have many granular itests or very long itests, but the nature of these testcases generally fits what we do in the liquidity edge cases itest

Reply:
I thought about that. But I think we have to be careful that we then don't run into false negatives. For example, we expect a payment to fail for a specific edge case reason, but it ends up failing because the mission control has recorded too many previous failures and gives up on the payment before even trying.
Sure, we could counter that by clearing mission control and/or checking for the correct failure.
But I think the extra overhead of 30 seconds per test is currently bearable in exchange for a clean test slate.

IMO we can do some cleanup and test case consolidation once we've released the current version.